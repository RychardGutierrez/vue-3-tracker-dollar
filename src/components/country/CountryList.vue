<template>
  <div v-for="tracker in listTrakers" :key="tracker.id">
    <CountryCard :country="tracker" @click="goToCountryByTracker(tracker)" />
  </div>
  <p v-if="listTrakers.length === 0" class="text-lg">
    No countries added. To start tracking a dollar to country, search in the
    field above.
  </p>
</template>

<script setup>
import useTrakerDollar from "@/composables/useTrakerDollar"
import useRouteCountry from '@/composables/useRouteCountry';

import CountryCard from './CountryCard.vue';

const { getTrakers, listTrakers } = useTrakerDollar();
const { goToCountryByTracker } = useRouteCountry()

getTrakers();
</script>

<style lang="scss" scoped>

</style>
