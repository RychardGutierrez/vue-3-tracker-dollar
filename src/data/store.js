import { uid } from 'uid';

export const updatedStore = ({
  country,
  official,
  symbol,
  savedDate,
  rate,
}) => {
  let data = [];
  try {
    if (localStorage.getItem('savedTrackers')) {
      data = JSON.parse(localStorage.getItem('savedTrackers'));
    }

    const trackerObj = {
      country,
      official,
      symbol,
      savedDate,
      rate,
      defaultRate: '1 USD',
      id: uid(),
    };

    data.push(trackerObj);
    localStorage.setItem('savedTrackers', JSON.stringify(data));
    return {
      id: trackerObj.id,
      error: false,
      message: 'Saved tracker',
    };
  } catch (error) {
    return {
      error: true,
      message: 'Error save tracker',
    };
  }
};

export const readStore = () => {
  let data = [];
  try {
    if (localStorage.getItem('savedTrackers')) {
      data = JSON.parse(localStorage.getItem('savedTrackers'));
    }

    return {
      data,
      error: false,
      message: 'Read tracker',
    };
  } catch (error) {
    return {
      error: true,
      message: 'Error read tracker',
    };
  }
};

export const removedStore = (idRemove) => {
  let data = [];
  try {
    if (localStorage.getItem('savedTrackers')) {
      data = JSON.parse(localStorage.getItem('savedTrackers'));
    }

    const updatedDate = data.filter(({ id }) => id !== idRemove);
    localStorage.setItem('savedTrackers', JSON.stringify(updatedDate));

    return {
      error: false,
      message: 'Removed tracker',
    };
  } catch (error) {
    return {
      error: true,
      message: 'Error removed tracker',
    };
  }
};
