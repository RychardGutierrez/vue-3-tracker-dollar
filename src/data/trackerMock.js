// https://api.exchangeratesapi.io/v1/timeseries? access_key = API_KEY& start_date = 2012-05-01& end_date = 2012-05-25
export const TRACKER_HISTORY = {
  success: true,
  timeseries: true,
  start_date: '2012-05-01',
  end_date: '2012-05-03',
  base: 'USD',
  rates: {
    '2022-09-01': {
      BOL: 6.9,
    },
    '2022-09-02': {
      BOL: 6.92,
    },
    '2022-09-03': {
      BOL: 6.98,
    },
  },
};

//  'https://api.apilayer.com/currency_data/convert?to=BOB&from=USD&amount=1',

export const CONVERTER_DOLLAR = {
  success: true,
  query: {
    from: 'USD',
    to: 'BOB',
    amount: 1,
  },
  info: {
    timestamp: 1664082963,
    quote: 6.95361,
  },
  result: 6.95361,
};

export const TRACKER_DOLLAR = {
  success: true,
  timeframe: true,
  start_date: '2022-09-18',
  end_date: '2022-09-25',
  source: 'USD',
  quotes: {
    '2022-09-19': {
      USDBOB: 6.828416,
    },
    '2022-09-20': {
      USDBOB: 6.90908,
    },
    '2022-09-21': {
      USDBOB: 6.915233,
    },
    '2022-09-22': {
      USDBOB: 6.914142,
    },
    '2022-09-23': {
      USDBOB: 6.90469,
    },
    '2022-09-24': {
      USDBOB: 6.95361,
    },
    '2022-09-25': {
      USDBOB: 6.95361,
    },
  },
};

export const COUNTRY_DATA = [
  {
    flags: {
      png: 'https://flagcdn.com/w320/bo.png',
      svg: 'https://flagcdn.com/bo.svg',
      alt: 'The flag of Bolivia is composed of three equal horizontal bands of red, yellow and green, with the national coat of arms centered in the yellow band.',
    },
    altSpellings: [
      'BO',
      'Buliwya',
      'Wuliwya',
      'Bolivia, Plurinational State of',
      'Plurinational State of Bolivia',
      'Estado Plurinacional de Bolivia',
      'Buliwya Mamallaqta',
      'Wuliwya Suyu',
      'Tetã Volívia',
    ],
  },
];

export const LIST_RESULT_SEARCH = {
  name: {
    common: 'Bolivia',
    official: 'Plurinational State of Bolivia',
    nativeName: {
      aym: {
        official: 'Wuliwya Suyu',
        common: 'Wuliwya',
      },
      grn: {
        official: 'Tetã Volívia',
        common: 'Volívia',
      },
      que: {
        official: 'Buliwya Mamallaqta',
        common: 'Buliwya',
      },
      spa: {
        official: 'Estado Plurinacional de Bolivia',
        common: 'Bolivia',
      },
    },
  },
  currencies: {
    BOB: {
      name: 'Bolivian boliviano',
      symbol: 'Bs.',
    },
  },
  altSpellings: [
    'BO',
    'Buliwya',
    'Wuliwya',
    'Bolivia, Plurinational State of',
    'Plurinational State of Bolivia',
    'Estado Plurinacional de Bolivia',
    'Buliwya Mamallaqta',
    'Wuliwya Suyu',
    'Tetã Volívia',
  ],
};

export const LIST_TRACKER_COUNTRY = [
  {
    country: 'bolivia',
    official: 'Plurinational State of Bolivia',
    symbol: 'BOB',
    savedDate: 'Feb 8, 2023',
    defaultRate: '1 USD',
    id: '9a9f07bea8f',
  },
  {
    country: 'unitedstatesminoroutlyingislands',
    official: 'United States Minor Outlying Islands',
    symbol: 'USD',
    savedDate: 'Feb 8, 2023',
    defaultRate: '1 USD',
    id: '6066f819001',
  },
];
