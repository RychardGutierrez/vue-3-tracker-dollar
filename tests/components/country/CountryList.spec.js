import { flushPromises, mount, shallowMount } from '@vue/test-utils';
import CountryList from '../../../src/components/country/CountryList.vue';

import { LIST_TRACKER_COUNTRY as mockListTrackerCountry } from '../../../src/data/trackerMock';

jest.mock('../../../src/composables/useTrakerDollar', () => {
  const useTrakerDollar = () => ({
    getTrakers: jest.fn(),
    listTrakers: mockListTrackerCountry,
  });

  return useTrakerDollar;
});

const mockGoToCountryByTracker = jest.fn();

jest.mock('../../../src/composables/useRouteCountry', () => {
  const useRouteCountry = () => ({
    goToCountryByTracker: mockGoToCountryByTracker,
  });

  return useRouteCountry;
});

describe('CountryList component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should to match with snapshot - mount', () => {
    const wraper = mount(CountryList);

    expect(wraper.html()).toMatchSnapshot();
  });

  test('should to match with snapshot - shallowMount', () => {
    const wraper = shallowMount(CountryList);

    expect(wraper.html()).toMatchSnapshot();
  });

  test('should call toCountryByTracker', async () => {
    const tracker = {
      country: 'bolivia',
      defaultRate: '1 USD',
      id: '9a9f07bea8f',
      official: 'Plurinational State of Bolivia',
      savedDate: 'Feb 8, 2023',
      symbol: 'BOB',
    };
    const wraper = shallowMount(CountryList);
    await flushPromises();

    let countryCart = wraper.find('country-card-stub');
    await countryCart.trigger('click');
    expect(mockGoToCountryByTracker).toHaveBeenCalledWith(tracker);

    countryCart = wraper.find('[data-test="card-9a9f07bea8f"]');
    await countryCart.trigger('click');
    expect(mockGoToCountryByTracker).toHaveBeenCalledWith(tracker);
  });
});
