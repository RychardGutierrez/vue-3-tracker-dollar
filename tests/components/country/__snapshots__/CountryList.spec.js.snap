// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryList component should to match with snapshot - mount 1`] = `
"<div>
  <div class=\\"flex py-4 px-4 bg-dollar-primary hover:bg-green-600 text-white rounded-md shadow-md cursor-pointer\\" data-test=\\"card-9a9f07bea8f\\">
    <div class=\\"flex flex-col flex-1\\">
      <h2 class=\\"text-2xl\\">Plurinational State of Bolivia</h2>
      <h3>Feb 8, 2023</h3>
    </div>
    <div class=\\"flex flex-col gap-2\\">
      <p class=\\"text-3cl self-end\\"> BOB</p>
      <p class=\\"text-3cl self-end\\">1 USD</p>
    </div>
  </div>
</div>
<div>
  <div class=\\"flex py-4 px-4 bg-dollar-primary hover:bg-green-600 text-white rounded-md shadow-md cursor-pointer\\" data-test=\\"card-6066f819001\\">
    <div class=\\"flex flex-col flex-1\\">
      <h2 class=\\"text-2xl\\">United States Minor Outlying Islands</h2>
      <h3>Feb 8, 2023</h3>
    </div>
    <div class=\\"flex flex-col gap-2\\">
      <p class=\\"text-3cl self-end\\"> USD</p>
      <p class=\\"text-3cl self-end\\">1 USD</p>
    </div>
  </div>
</div>
<!--v-if-->"
`;

exports[`CountryList component should to match with snapshot - shallowMount 1`] = `
"<div>
  <country-card-stub country=\\"[object Object]\\" data-test=\\"card-9a9f07bea8f\\"></country-card-stub>
</div>
<div>
  <country-card-stub country=\\"[object Object]\\" data-test=\\"card-6066f819001\\"></country-card-stub>
</div>
<!--v-if-->"
`;
